File: problems/mandelbrot_set/cpp/mandelbrot.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <complex>
#include <thread>
#include <chrono>

// Image and Mandelbrot parameters
const int WIDTH = 1600;
const int HEIGHT = 1200;
const double ZOOM = 300.0;
const int MAX_ITER = 1000;
const int NUM_THREADS = 8; // Number of worker threads for parallelism

// Mandelbrot iteration function: returns number of iterations before escape
int mandelbrot(double x, double y) {
    std::complex<double> z(0.0, 0.0);     // Initial z value
    std::complex<double> c(x, y);         // Constant c = current point
    int iter = 0;

    // Iterate until |z|^2 > 4 or max iterations reached
    while (std::norm(z) <= 4.0 && iter < MAX_ITER) {
        z = z * z + c;  // Mandelbrot formula: z = z^2 + c
        ++iter;
    }
    return iter; // Return iteration count (used for color mapping)
}

// Thread worker: computes rows [start, end) and stores in 'result'
void compute_rows(int start, int end, std::vector<std::vector<int>>& result) {
    for (int py = start; py < end; ++py) {
        std::vector<int> row(WIDTH); // Create row buffer
        for (int px = 0; px < WIDTH; ++px) {
            // Convert pixel to complex plane coordinates
            double x = static_cast<double>(px) / ZOOM - 2.0;
            double y = static_cast<double>(py) / ZOOM - 1.5;
            row[px] = mandelbrot(x, y); // Compute iteration count
        }
        result[py] = std::move(row); // Store row in result vector
    }
}

int main() {
    std::vector<std::vector<int>> result(HEIGHT); // Stores the full image data
    std::vector<std::thread> threads;             // Container for thread objects

    auto start_time = std::chrono::high_resolution_clock::now(); // Start timing

    // Divide work among threads (each thread processes a block of rows)
    int rows_per_thread = HEIGHT / NUM_THREADS;
    for (int i = 0; i < NUM_THREADS; ++i) {
        int start = i * rows_per_thread;
        int end = (i == NUM_THREADS - 1) ? HEIGHT : start + rows_per_thread;
        threads.emplace_back(compute_rows, start, end, std::ref(result)); // Launch thread
    }

    // Wait for all threads to complete
    for (auto& t : threads) {
        t.join();
    }

    auto end_time = std::chrono::high_resolution_clock::now(); // End timing
    std::chrono::duration<double> elapsed = end_time - start_time;
    std::cout << "Computation Time: " << elapsed.count() << " seconds\n";

    // Write the output to a CSV file
    std::ofstream out("mandelbrot_output.csv");
    for (const auto& row : result) {
        for (size_t i = 0; i < row.size(); ++i) {
            out << row[i] << (i + 1 == row.size() ? "\n" : ","); // Format CSV
        }
    }
    out.close();
    std::cout << "Output saved to mandelbrot_output.csv\n";

    return 0;
}

File: problems/mandelbrot_set/ocaml/mandelbrot.ml
  open Domainslib
(* Define image dimensions and computation parameters *)
  let width = 1600
  let height = 1200
  let zoom = 300.0
  let max_iter = 1000
  
  (* Higher-order function: takes a pixel coordinate -> returns iterations *)
  let mandelbrot_iter x y =
    let rec loop zx zy i =
      if i >= max_iter then i
      else
        let zx2 = zx *. zx in
        let zy2 = zy *. zy in
        let sum = zx2 +. zy2 in
        if sum > 4.0 then i     (* Escapes if magnitude exceeds threshold *)

        (* Recurse with new z values following the Mandelbrot formula *)
        else loop (zx2 -. zy2 +. x) (2.0 *. zx *. zy +. y) (i + 1)
    in
    loop 0.0 0.0 0
  
  (* Lazily computes a row of the Mandelbrot set image.
   Each pixel's iteration count is computed and stored in an array. *)
  let compute_row row =
    Lazy.force (
      Lazy.from_fun (fun () ->
        Array.init width (fun col ->
          let x = (float_of_int col /. zoom) -. 2.0 in
          let y = (float_of_int row /. zoom) -. 1.5 in
          mandelbrot_iter x y))
    )
  
    let () =
    (* Create a pool of 4 domains for parallel computation *)
    (* 4 was found to be the ideal size after contrasting with 2, 8, 12 domains *)
    let pool = Task.setup_pool ~num_domains:4 () in
    let t0 = Unix.gettimeofday () in
    (* Parallel computation of all rows using Domainslib task pool *)
    let output = Task.run pool (fun () ->
      Array.init height (fun row ->
        Task.async pool (fun () -> compute_row row)
      )
      |> Array.map (Task.await pool)
    ) in
    let t1 = Unix.gettimeofday () in
    (* Clean up the domain pool *)
    Task.teardown_pool pool;
  
    Printf.printf "Computation Time: %.4f seconds\n" (t1 -. t0);
  
    (* Write output *)
    let oc = open_out "mandelbrot_output.csv" in
    Array.iter (fun row ->
      Array.iteri (fun i iter ->
        Printf.fprintf oc "%d%s" iter (if i = width - 1 then "\n" else ",")
      ) row
    ) output;
    close_out oc;
  
    print_endline "Parallel OCaml Mandelbrot set written to mandelbrot_output.csv"


